src/fido2_child/fido2_child_credentials.c

```diff
@@ -0,0 +1,291 @@
+
+    ret = fido_cred_set_user(cred, userid, sizeof(userid), data->shortname,
+                             NULL, NULL);
+    if (ret != FIDO_OK) {
+        DEBUG(SSSDBG_OP_FAILURE, "fido_cred_set_user failed [%d]: %s.\n",
+              ret, fido_strerr(ret));
+        goto done;
+    }
+
+    ret = fido_cred_set_rk(cred, FIDO_OPT_FALSE);
+    if (ret != FIDO_OK) {
+        DEBUG(SSSDBG_OP_FAILURE, "fido_cred_set_rk failed [%d]: %s.\n",
+              ret, fido_strerr(ret));
+        goto done;
+    }
+
+    ret = fido_cred_set_uv(cred, data->user_verification);
+    if (ret != FIDO_OK) {
+        DEBUG(SSSDBG_OP_FAILURE, "fido_cred_set_uv failed [%d]: %s.\n",
+              ret, fido_strerr(ret));
+        goto done;
+    }
+
+done:
+    return ret;
+}
+
+ssize_t
+read_pin(char **pin)
+{
+    char *line_ptr = NULL;
+    struct termios old, new;
+    size_t line_len = 0;
+    ssize_t bytes_read;
+    ssize_t ret;
+
+    ret = tcgetattr(STDIN_FILENO, &old);
+    if (ret != 0) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "Unable get the parameters associated with stdin.\n");
+        goto done;
+    }
+    new = old;
+    new.c_lflag &= ~ECHO;
+    ret = tcsetattr(STDIN_FILENO, TCSAFLUSH, &new);
+    if (ret != 0) {
+        DEBUG(SSSDBG_OP_FAILURE, "Unable to turn echoing off.\n");
+        goto done;
+    }
+
+    printf("Enter PIN: ");
+    bytes_read = getline(&line_ptr, &line_len, stdin);
+    if (bytes_read == -1) {
+        DEBUG(SSSDBG_OP_FAILURE, "getline failed [%d]: %s.\n",
+              errno, sss_strerror(errno));
+    } else {
+        // Remove the end of line '\n' character
+        line_ptr[--bytes_read] = '\0';
+    }
+    printf("\n");
+
+    ret = tcsetattr(STDIN_FILENO, TCSAFLUSH, &old);
+    if (ret != 0) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "Unable to restore parameters associated with stdin.\n");
+        goto done;
+    }
+
+    ret = bytes_read;
+    *pin = line_ptr;
+
+done:
+    return ret;
+}
+
+errno_t
+generate_credentials(struct fido2_data *data, fido_dev_t *dev,
+                     fido_cred_t *cred)
+{
+    char *pin = NULL;
+    ssize_t pin_len;
+    bool pin_enabled;
+    errno_t ret;
+
+    pin_enabled = fido_dev_has_pin(dev);
+    if (data->user_verification == FIDO_OPT_FALSE && pin_enabled == true) {
+        DEBUG(SSSDBG_CONF_SETTINGS, "Policy disabled user verification but "
+              "the key enforces it. Thus, enabling the user verification "
+              "feature to avoid authentication errors.\n");
+        data->user_verification = FIDO_OPT_TRUE;
+    }
+
+    if (data->user_verification == FIDO_OPT_FALSE) {
+        ret = fido_dev_make_cred(dev, cred, NULL);
+    } else {
+        pin_len = read_pin(&pin);
+        if (pin_len == -1) {
+            return ERR_INPUT_PARSE;
+        }
+
+        ret = fido_dev_make_cred(dev, cred, pin);
+        sss_erase_mem_securely(pin, pin_len);
+        free(pin);
+    }
```

Comment:
> Is the intention here to always rely on `clientPin` for user verification? If so, there is no need for the `fido_cred_set_uv()` call on line 99 since that is only used to toggle a request for some form of built-in UV (which `clientPin` is not).

No, we should allow other ways of verifying a user.

Can you specify some other way of verifying a user? I guess the fingerprint, but is there any other?

> Note that there may also be authenticators out there that do not support `clientPin` (the specification does not mandate support for it), in which case they may fail in the `fido_dev_make_cred()` call on line 182. (These type of authenticators do seem to be rare though.)

What would `fido_dev_has_pin()` return in that case?

> Additional logic will be required here to handle both, an example could be to ask the user to provide a PIN if `fido_dev_has_pin()` is true, request built-in UV if `fido_dev_has_uv()` is true, and neither otherwise -- depending on your preferred user experience.

The proposal seems good but it delegates the organization policy to the key itself. From the FIDO2 specification is it mandatory to provide at least a user verification method?

Reply:
> Can you specify some other way of verifying a user? I guess the fingerprint, but is there any other?

In terms of the specification, there are two ways of providing user verification

1. [`clientPin`](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-20210615.html#getinfo-clientpin), i.e. the authenticator is capable of accepting a PIN from the client (in this case, sssd)
2. [built-in UV](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-20210615.html#built-in-user-verification-method), i.e. the authenticator supports built-in on-device user verification. This can be either biometrics, a secure keypad built directly into the authenticator, or something else.

> What would `fido_dev_has_pin()` return in that case?

Both `fido_dev_has_pin()` and `fido_dev_supports_pin()` would return false.

> From the FIDO2 specification is it mandatory to provide at least a user verification method?

No, authenticators may choose to not implement any type of UV.